### PR TITLE
Issue #148 - Implement SequenceEqual operator

### PR DESCRIPTION
**Pull request by MarekLabuz**
PR implements SequenceEqual operator requested in #148 
It includes operator code, tests and example.

### CODE REVIEW

**Comment by MarekLabuz** [comment]
I have closed the pull request and reopened it due to the failing tests of go1.9 (other versions passed) - the second attempt was successful as stated above. It turned out that TestStartOperator failed, so I have inspected it. There are two mock network request that are executed with 5 milliseconds delay between them and the test itself relays on the order of their execution. I guess it may cause a race condition and this is what happened in my case. Shall we increase the delay to minimise the probability of overlapping executions?

**Comment by teivah** [comment]
> Shall we increase the delay to minimise the probability of overlapping executions?

Thanks, it's now part of the `v2` branch (https://github.com/ReactiveX/RxGo/pull/171)

File: observable.go [hunk 1]
```diff
@@ -51,6 +51,7 @@ type Observable interface {
 	Reduce(apply Function2) OptionalSingle
 	Repeat(count int64, frequency Duration) Observable
 	Scan(apply Function2) Observable
+	SequenceEqual(obs Observable) Observable
```
Comment (teivah) [inline observable.go]: Could you please rather return a `Single`?

File: observable.go [hunk 2]
```diff
@@ -413,6 +414,74 @@ func (o *observable) Scan(apply Function2) Observable {
...
+		s1, sequence1 := inputSequence1[0], inputSequence1[1:]
+		s2, sequence2 := inputSequence2[0], inputSequence2[1:]
+		return s1 == s2, sequence1, sequence2
+	}
+	return true, inputSequence1, inputSequence2
+}
+
+// SequenceEqual emits true if an Observable and the input Observable emit the same items.
+// When any Observable emits different item, it emits false.
+func (o *observable) SequenceEqual(obs Observable) Observable {
```
Comment (teivah) [inline observable.go]: Since a recent merge in the `v2`, there's another way of creating operators (the reason is for better reuse across different operator types). Your code doesn't compile anymore. Can you please check other operators and follow the same logic?

**Comment by MarekLabuz** [comment]
I've made the changes @teivah , however build failed only on Go 1.10 due to the coveralls' error: 
```
Bad response status from coveralls: 422
{"message":"Couldn't find a repository matching this job.","error":true}
```

File: observable_test.go [hunk 3]
```diff
@@ -618,6 +618,29 @@ func TestObservableDistinctUntilChanged(t *testing.T) {
 	assert.Exactly(t, []int{1, 2, 1, 3}, nums)
 }
 
+func TestObservableSequenceEqualWithCorrectSequence(t *testing.T) {
+	sequence := Just(2, 5, 12, 43, 98, 100, 213)
+	result := Just(2, 5, 12, 43, 98, 100, 213).SequenceEqual(sequence)
+	AssertThatSingle(t, result, HasItems(true))
```
Comment (teivah) [inline observable_test.go]: `HasValue` is the assertion to be used with `Single`

File: observable_test.go [hunk 4]
```diff
@@ -618,6 +618,29 @@ func TestObservableDistinctUntilChanged(t *testing.T) {
...
+}
+
+func TestObservableSequenceEqualWithIncorrectSequence(t *testing.T) {
+	sequence := Just(2, 5, 12, 43, 98, 100, 213)
+	result := Just(2, 5, 12, 43, 15, 100, 213).SequenceEqual(sequence)
+	AssertThatSingle(t, result, HasItems(false))
```
Comment (teivah) [inline observable_test.go]: Same remark

File: observable_test.go [hunk 5]
```diff
@@ -618,6 +618,29 @@ func TestObservableDistinctUntilChanged(t *testing.T) {
...
+}
+
+func TestObservableSequenceEqualWithDifferentLengthSequence(t *testing.T) {
+	sequenceShorter := Just(2, 5, 12, 43, 98, 100)
+	sequenceLonger := Just(2, 5, 12, 43, 98, 100, 213, 512)
+
+	resultForShorter := Just(2, 5, 12, 43, 98, 100, 213).SequenceEqual(sequenceShorter)
+	AssertThatSingle(t, resultForShorter, HasItems(false))
```
Comment (teivah) [inline observable_test.go]: Same remark

File: observable_test.go [hunk 6]
```diff
@@ -618,6 +618,29 @@ func TestObservableDistinctUntilChanged(t *testing.T) {
...
+
+	resultForLonger := Just(2, 5, 12, 43, 98, 100, 213).SequenceEqual(sequenceLonger)
+	AssertThatSingle(t, resultForLonger, HasItems(false))
```
Comment (teivah) [inline observable_test.go]: Same remark

File: observable_test.go [hunk 7]
```diff
@@ -618,6 +618,29 @@ func TestObservableDistinctUntilChanged(t *testing.T) {
 	assert.Exactly(t, []int{1, 2, 1, 3}, nums)
 }
```
Comment (teivah) [inline observable_test.go]: Could you also please one test with empty observable? Something like this:

```go
func TestObservableSequenceEqualWithEmpty(t *testing.T) {
	result := Empty().SequenceEqual(Empty())
	AssertThatSingle(t, result, HasValue(true))
}
```

File: observable_test.go (hunk 3 above)
Comment (teivah) [reply observable_test.go]: For the time being the assertion API is not typed, this might be improved at some point.

**Comment by teivah** [comment]
> I've made the changes @teivah , however build failed only on Go 1.10 due to the coveralls' error

That was a transient error, the build is now okay.

**Comment by MarekLabuz** [comment]
Thank you for your feedback @teivah , I've just pushed the changes.
